src/unpack.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use libc::timeval;
+use std::collections::HashMap;
+use std::ffi::CString;
+use std::fs;
+use std::io;
+use std::path::Path;
+use tar::Archive;
+
+/// Unpack the contents of tarball to the destination path
+pub fn unpack(input: Vec<u8>, destination: &Path) -> Result<()> {
+    let mut archive = Archive::new(input.as_slice());
+
+    if destination.exists() {
+        return Err(anyhow!(
+            "unpack destination {:?} already exists",
+            destination
+        ));
+    }
+
+    fs::create_dir_all(destination)?;
+
+    let mut dirs: HashMap<CString, [timeval; 2]> = HashMap::default();
+    for file in archive.entries()? {
+        let mut file = file?;
+        file.unpack_in(destination)?;
+
+        // tar-rs crate only preserve timestamps of files,
+        // symlink file and directory are not covered.
+        if file.header().entry_type().is_symlink() || file.header().entry_type().is_dir() {
+            let mtime = file.header().mtime()? as i64;
+
+            let atime = timeval {
+                tv_sec: mtime,
+                tv_usec: 0,
+            };
+            let path = CString::new(format!(
+                "{}/{}",
+                destination.display(),
+                file.path()?.display()
+            ))?;
+
+            let times = [atime, atime];
+
+            if file.header().entry_type().is_dir() {
+                dirs.insert(path, times);
+            } else {
+                let ret = unsafe { libc::lutimes(path.as_ptr(), times.as_ptr()) };
+                if ret != 0 {
+                    return Err(anyhow!(
+                        "change symlink file: {:?} utime error: {:?}",
+                        path,
+                        io::Error::last_os_error()
+                    ));
+                }
```

Comment:
1. This code appears also below; consider extracting it into a function (set_file_timestamps ?)
2. Why do we bother changing the atime? Wouldn't it be more "correct" to leave it as-is (= "now")? (asking because I'm not sure about this)
3. filetime crate has set_symlink_file_times. Why don't we use it?

Reply:
> 1. This code appears also below; consider extracting it into a function (set_file_timestamps ?)

we have two set time operations one is utime and another is lutime for symlink and each only call once, it make sense to extract into a function if it used at multiple places.

> 2. Why do we bother changing the atime? Wouldn't it be more "correct" to leave it as-is (= "now")? (asking because I'm not sure about this)

For container image, we need keep it consistent with the state when the image is packed, this is the common way for the image service.

> 3. filetime crate has set_symlink_file_times. Why don't we use it?

Yes, we can use filetime, the reason is to simplify the dependency,  maybe we can swith to filetime crate when we start working on support different platforms.

Reply:
> > 1. This code appears also below; consider extracting it into a function (set_file_timestamps ?)
> 
> we have two set time operations one is utime and another is lutime for symlink and each only call once, it make sense to extract into a function if it used at multiple places.

You're right, I missed that. I think that `lutimes` should work for both cases. But the existing code is fine anyway.

> 
> > 2. Why do we bother changing the atime? Wouldn't it be more "correct" to leave it as-is (= "now")? (asking because I'm not sure about this)
> 
> For container image, we need keep it consistent with the state when the image is packed, this is the common way for the image service.

OK.

> 
> > 3. filetime crate has set_symlink_file_times. Why don't we use it?
> 
> Yes, we can use filetime, the reason is to simplify the dependency, maybe we can swith to filetime crate when we start working on support different platforms.

Yes you're right, if you only need these two functions then no need to depend on the filetime crate.